### PR TITLE
[tlse] tls for GlanceAPI pod configuration

### PR DESCRIPTION
**Pull request by stuggi**
Public/Internal service cert secrets and the CA bundle secret can be passed to configure httpd virtual hosts for tls termination. The certs get direct mounted to the appropriate place in etc/pki/tls/certs/%s.crt|key and a CA bundle to
/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem . Job deployments for bootstrap/cron get the CA bundle added if configured.

Also indexes the named input resources for password, CA bundle, and endpoint secrets to be able to watch them for a change and reconcile.

Depends-On: https://github.com/openstack-k8s-operators/lib-common/pull/428

Jira: [OSPRH-1233](https://issues.redhat.com//browse/OSPRH-1233)

### CODE REVIEW

**Comment by stuggi** [comment]
needs a rebase for the glanceapi change

File: config/samples/glance_v1beta1_glanceapi.yaml [hunk 1]
```diff
@@ -19,3 +19,11 @@ spec:
   replicas: 1
   storageRequest: 10G
   secret: glance-secret
+  #tls:
```
Comment (fmount) [inline config/samples/glance_v1beta1_glanceapi.yaml:22]: Do we want to provide an example or a dedicated section in the `README` ? I think this is going to be a common pattern for all the operators, but having a bit of doc regarding these new parameters would help people understand how the feature works.

File: controllers/glanceapi_controller.go [hunk 2]
```diff
@@ -178,6 +184,53 @@ func (r *GlanceAPIReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *GlanceAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// index passwordSecretField
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &glancev1.GlanceAPI{}, passwordSecretField, func(rawObj client.Object) []string {
```
Comment (fmount) [inline controllers/glanceapi_controller.go:188]: This part could be partially optimized by iterating over `glanceAPIWatchFields` and index the current field, but I'm not sure how the resulting  `cr.Spec.<field>` can be dynamically resolved.

**Review by fmount — COMMENTED** [review]
I'm +1 in general, I don't see any blocking thing that prevent this change to land. I had just a few minor comments related to code optimization that I'm not even sure we can easily achieve, but dedup can also represent something we can do in a follow up.

File: config/samples/glance_v1beta1_glanceapi.yaml (hunk 1 above)
Comment (stuggi) [reply config/samples/glance_v1beta1_glanceapi.yaml:22]: I have started this PR for an addition on internal certs https://github.com/stuggi/docs/pull/new/tls_internal . should we still have something dedicated in the `README` ? just a link to the global conf. What kind of example you have in mind? a glance CR ?

File: controllers/glanceapi_controller.go [hunk 3]
```diff
@@ -471,9 +551,11 @@ func (r *GlanceAPIReconciler) reconcileNormal(ctx context.Context, instance *gla
 	// queries by label
 
 	serviceLabels := map[string]string{
-		common.AppSelector:       glance.ServiceName,
+		common.AppSelector: glance.ServiceName,
+		//common.AppSelector:       fmt.Sprintf("%s-%s", glance.ServiceName, instance.Spec.APIType),
```
[the remark this answers is not in the document]
Comment (stuggi) [reply controllers/glanceapi_controller.go]: done

File: config/samples/glance_v1beta1_glanceapi.yaml (hunk 1 above)
Comment (fmount) [reply config/samples/glance_v1beta1_glanceapi.yaml:22]: ah ok, I think having a general doc in that repo is better than having yet another dedicate section here where we explain the fields that are part of the new API definition.
Considering we're going to drive this process from the openstack-operator, I think we're good here. I'll take care about any extra doc in a follow up.

File: controllers/glance_common.go [hunk 4]
```diff
@@ -28,7 +32,26 @@ import (
...
+// fields to index to reconcile when change
+const (
+	passwordSecretField     = ".spec.secret"
+	caBundleSecretNameField = ".spec.tls.caBundleSecretName"
+	tlsAPIInternalField     = ".spec.tls.api.internal.secretName"
+	tlsAPIPublicField       = ".spec.tls.api.public.secretName"
+)
+
+var (
+	glanceWatchFields = []string{
```
[the remark this answers is not in the document]
Comment (fmount) [reply controllers/glance_common.go:46]: it's ok to keep it as it is

File: controllers/glance_controller.go [hunk 5]
```diff
@@ -200,9 +218,47 @@ func (r *GlanceReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 		Owns(&rbacv1.RoleBinding{}).
+		Watches(
+			&source.Kind{Type: &corev1.Secret{}},
+			handler.EnqueueRequestsFromMapFunc(r.findObjectsForSrc),
+			builder.WithPredicates(predicate.ResourceVersionChangedPredicate{}),
+		).
 		Complete(r)
 }
 
+func (r *GlanceReconciler) findObjectsForSrc(src client.Object) []reconcile.Request {
```
[the remark this answers is not in the document]
Comment (fmount) [reply controllers/glance_controller.go:229]: I imagine you can pass them to the `findObjectsForSrc` invocation, I'm just not sure how it reacts having already a `client.Object` as default parameter.

[the remark this answers is not in the document]
Comment (fmount) [reply controllers/glance_controller.go:229]: So let's keep it as it is as I think we need to experiment a bit on that front

[the remark this answers is not in the document]
Comment (stuggi) [reply controllers/glance_controller.go:229]: sounds good

File: config/samples/glance_v1beta1_glanceapi.yaml (hunk 1 above)
Comment (stuggi) [reply config/samples/glance_v1beta1_glanceapi.yaml:22]: nice, thanks

**Comment by stuggi** [comment]
error: build error: Failed to push image: trying to reuse blob sha256:07a64a71e01156f8f99039bc246149925c6d1480d3957de78510bbec6ec68f7a at destination: pinging container registry quay.rdoproject.org: Get "https://quay.rdoproject.org/v2/": net/http: TLS handshake timeout

**Comment by stuggi** [comment]
error: build error: Failed to push image: trying to reuse blob sha256:07a64a71e01156f8f99039bc246149925c6d1480d3957de78510bbec6ec68f7a at destination: pinging container registry quay.rdoproject.org: Get "https://quay.rdoproject.org/v2/": dial tcp 38.129.56.158:443: i/o timeout

**Comment by stuggi** [comment]
logger.go:42: 16:21:34 | glance_split/3-scale-down-glanceapis | Error from server (InternalError): Internal error occurred: failed calling webhook "mglance.kb.io": failed to call webhook: Post "[https://glance-operator-controller-manager-service.openstack-operators.svc:443/mutate-glance-openstack-org-v1beta1-glance?timeout=10s](https://glance-operator-controller-manager-service.openstack-operators.svc/mutate-glance-openstack-org-v1beta1-glance?timeout=10s)": no endpoints available for service "glance-operator-controller-manager-service"

**Review by fmount — APPROVED** [review]
/lgtm

**Pull request merged**